### PR TITLE
[kuduraft] [proxy] auto generate proxy topology using raft configs

### PR DESCRIPTION
**Pull request by bhatvinay**
Summary:
Manual updation of proxy topology (by external automation) needs proper
coordination across different systems as and when new peers are added,
removed or modified. Failure handling becomes tricky and difficult. So, it was
decided that it is better to  generate 'proxy topology' and routing
tables based on simple rules through raft configs. The routing rules
gets updated automatically when new config are proposed by the leader. This PR
tries to implement the same.

Different routing policies are defined in proxy_policy.h. There are two
policies that are defined and implemented:
1. Durable routing policy - existing policy where the routing topology
   is explicitly specified by external entities.
2. Simple region based routing policy - A new rule based proxy routing. It
   tries to keep parity with current vanilla deployment. At a high level,
   the rules used are:
 a) Each region has at most one designated 'proxy_peer' - it is chosen to be the
    peer which has a backing database. Leader ships messages to this
    'proxy_peer' when it has to send messages to any other peer in this
    region.
 b) Some regions may not have a 'proxy_peer'. Leader ships messages
    directly to all peers in such regions.
 c) There could be multiple 'candidate' proxy_peers in a given region - only
    one is chosen to be the real 'proxy_peer' for the region. This
    should not be required when region splitting is eventually
    implemented.
 d) When a new peer is added, we do not change any existing peer's proxy
    routing. This is to avoid churns in stable proxy routes on unrelated
    config changes

Implementation Notes:
1. An interface is defined through "IRoutingTable". The interface needs to be
implemented to support various routing policies.
2. A container holding all the roting policies is intantiated during
startup.
3. The policy can be shifted dynamically.

The container method was used to avoid additional locking overhead while
also supporting dynamic changes to proxy policy. The default policy is retained
to be 'DurableRoutingPolicy' to make it compatible with existing
implementation.

As the name suggests, DurableRoutingTable is persisted in cmeta - hence it is
continued to be instantiated and owned by cmeta_manager.
SimpleRegionRoutingTable is not persisted and is rebuilt using configs
on-the-fly. Hence it need not be managed by cmeta_manager.

Rules for SimpleRegionRoutingPolicy should get us a long way into
production deployment. In future, we could implement new policy with
rules for 'region-proying-another-region' to reduce cross region traffic (for
example cross atlantic traffic). If an external entity needs to implant (and
override) a topology, then the policy needs to be changed.
SimpleRegionRoutingPolicy is not persisted and hence it does not make sense to
allow external entities to provide/override arbitrary routing topology.

Test Plan: Tests in fbcode

Reviewers: arahut, mpercy

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: src/kudu/consensus/proxy_policy.h
Comment (mpercy) [inline src/kudu/consensus/proxy_policy.h:29]: This is fine, but we will need to extend this to handle multiple potential proxy peers in the future.

File: src/kudu/consensus/routing.h
Comment (mpercy) [inline src/kudu/consensus/routing.h]: nit: no need for explicit if there are no args to the constructor

File: src/kudu/consensus/routing.cc
Comment (mpercy) [inline src/kudu/consensus/routing.cc]: nit: how about `CreateRoutingTable()` since `RoutingTableFactory` looks like a class

File: src/kudu/consensus/routing.cc
Comment (mpercy) [inline src/kudu/consensus/routing.cc:558]: It would be nice to be able to support multi-hop in the future, but you're right that we don't need it right now.

File: src/kudu/consensus/routing.cc
Comment (mpercy) [inline src/kudu/consensus/routing.cc]: nit: consider renaming this map (and the class member) to something like `dst_to_next_hop_map`

File: src/kudu/consensus/routing.cc
Comment (mpercy) [inline src/kudu/consensus/routing.cc]: Why bother creating a self-loop edge? In the existing impl this would be `emplace(peer.uuid, leader.uuid)` which makes more sense to me (but as I am writing this comment I'm still reading through the impl trying to understand it).

**Review by mpercy — APPROVED** [review]
LGTM so far. Have you thought yet about how we will plumb in the additional metadata we will need in the future, like subregions or proxying across multiple regions? I don't think we have to figure that out to merge this, though, assuming you've been able to test this.

File: src/kudu/consensus/proxy_policy.h
Comment (bhatvinay) [reply src/kudu/consensus/proxy_policy.h:29]: Yes, have some ideas - but not needed immediately. For example,  we can have a list of 'valid proxy peers' and simply round-robin between them.

File: src/kudu/consensus/routing.cc
Comment (bhatvinay) [reply src/kudu/consensus/routing.cc]: Agree. ProxyTopologyPB is not used in this simple routing policy and is maintained only for reporting purpose. Will see if I can change this

**Comment by bhatvinay** [comment]
> LGTM so far. Have you thought yet about how we will plumb in the additional metadata we will need in the future, like subregions or proxying across multiple regions? I don't think we have to figure that out to merge this, though, assuming you've been able to test this.

For subregions, I will leave it to a later time to decide how it is handled. One way is to simply add a suffix to the region to differentiate - say prn1, prn2, prn3 etc. Another way is to add an explicit attribute in PeerAttrsPB as `peer_sub_region` and change it based on that. Not sure what is being planned.

For proxying across regions, we could provide a map of preferred 'proxy_regions' and use it to instantiate a new routing policy. Something like, {(cln, lla), (eag, prn)} == 'lla' acts as a proxy-region for 'cln', 'prn' acts as a proxy region for 'eag'.

Not yet ready for merge. I have to figure out how to allow overriding the default built routing tables. Have written some basic unit/integration tests in fbcode, but needs more work. This was an early upload to gather some feedback!

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc]: Is the extra "FromLeader" required in this function name? It seems to me more confusing as I can't understand how the Leader is coming into play here.

Comment (bhatvinay) [reply src/kudu/consensus/consensus_queue.cc]: "PeerMessageQueue::GetNextRoutingHopFromLeader()" is only called on the leader (to get the next hop along the routing tree). Hence the name (suggesting next hop to go from the leader). I can change it if it no longer makes sense.

File: src/kudu/consensus/routing.cc
Comment (bhatvinay) [reply src/kudu/consensus/routing.cc]: Fixed it by not creating any entry. So, if 'nexthop' does not exist, the deafult is to route directly

File: src/kudu/consensus/routing.h
Comment (bhatvinay) [reply src/kudu/consensus/routing.h]: Will fix this

File: src/kudu/consensus/routing.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/routing.cc]: wondering if we should sort the peers to maintain a proper order in regions where there are more
instances with backing dbs. this will ensure that the peer selection is stable.

Ignore this comment if raft_config peers always returns the same order of peers.

File: src/kudu/consensus/raft_consensus.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:4220]: seems to me the only place where we update routing_table_ is here and in the Init function of RaftConsensus. If we instead had a construct like an atomic shared pointer we could have delegated
all the locking to the Routing tree itself?

File: src/kudu/consensus/routing.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/routing.cc:652]: Where are the UpdateRaftConfig or RebuildProxyTopology functions being called?
Has this been integrated into Config change or is this in a later diff?

Is there a possibility that in an add operation the recently added member becomes the proxy
node?

File: src/kudu/consensus/raft_consensus.cc
Comment (bhatvinay) [reply src/kudu/consensus/raft_consensus.cc:4220]: Discussed offline. We have to think through this locking and ways to get rid of it. One approach is to have both durable and simple routing table inside a container and only use one lock internally (so that we never update the pointer to this external container and hence no locking will be needed here). Will do that in a subsequent upload (either this PR or the next one)

File: src/kudu/consensus/routing.cc
Comment (bhatvinay) [reply src/kudu/consensus/routing.cc:652]: Discussed offline. Routing table's UpdateRaftConfig() is called whenever a new config is proposed/committed/aborted. Hence the topology gets updated there automatically.

**Review by anirbanr-fb — APPROVED** [review]

**Pull request merged**
